src/number_theory/arithmetic_function.lean

```diff
@@ -0,0 +1,228 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+import algebra.big_operators.ring
+import number_theory.divisors
+
+/-!
+# Arithmetic Functions and Dirichlet Convolution
+
+This file defines arithmetic functions, which are functions from `ℕ` to a specified type that map 0
+to 0. In the literature, they are often instead defined as functions from `ℕ+`. These arithmetic
+functions are endowed with a multiplication, given by Dirichlet convolution, and pointwise addition,
+to form the Dirichlet ring.
+
+## Main Definitions
+ * `arithmetic_function α` consists of functions `f : ℕ → α` such that `f 0 = 0`.
+
+## Tags
+arithmetic functions, dirichlet convolution, divisors
+
+-/
+
+open finset
+open_locale big_operators
+
+namespace nat
+variable (α : Type*)
+
+/-- An arithmetic function is a function from `ℕ` that maps 0 to 0. In the literature, they are
+  often instead defined as functions from `ℕ+`. Multiplication on `arithmetic_functions` is by
+  Dirichlet convolution. -/
+structure arithmetic_function [has_zero α] :=
+(to_fun : ℕ → α)
+(map_zero' : to_fun 0 = 0)
+
+variable {α}
+
+namespace arithmetic_function
+
+section has_zero
+variable [has_zero α]
+
+instance : has_coe_to_fun (arithmetic_function α) := ⟨λ _, ℕ → α, to_fun⟩
+
+@[simp] lemma to_fun_eq (f : arithmetic_function α) : f.to_fun = f := rfl
+
+@[simp]
+lemma map_zero {f : arithmetic_function α} : f 0 = 0 := f.map_zero'
+
+theorem coe_inj ⦃f g : arithmetic_function α⦄ (h : (f : ℕ → α) = g) : f = g :=
```

Comment:
Could you write this as iff, or using `function.injective`?

Reply:
I can make it an iff. It looks like the other lemmas called `coe_inj` skew about 50-50 between this format (copied from `ring_hom.coe_inj`) and the same but an iff.